torch/_functorch/_aot_autograd/traced_function_transforms.py

```diff
@@ -587,6 +586,21 @@ def metadata_fn(*primals):
     )
 
 
+class PropagateUnbackedSymInts(torch.fx.Interpreter):
+    def run_node(self, n: torch.fx.Node):
+        import sympy
+
+        result = super().run_node(n)
+        # TODO: handle Tensor returns
```

Comment:
for my understanding: I would naively thought that:

(1) for an fx node that produces a tensor `x` with unbacked symints, there will be a later fx node that grabs that unbacked symint directly before anyone uses it (`aten.sym_size(x, dim)`)

(2) therefore we should be guaranteed to do the replacement / equality check later, before the symint is used anywhere

Is (1) not guaranteed?

Reply:
It is not guaranteed! I actually had homework to come up with a test case that triggers this. Here it is:

```
    @torch._dynamo.config.patch(
        capture_dynamic_output_shape_ops=True
    )
    def test_nonzero_no_realloc(self, device):
        @torch.compile(fullgraph=True)
        def f(x, y):
            z = x.nonzero()
            return torch.split(z, [y.size(0)])

        f(torch.tensor([1, 0, 1, 1, 0, 1, 0]), torch.randn(4))
```

This gives

```
[2024-01-25 08:34:32,887] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] create_env
[2024-01-25 08:34:32,895] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] create_unbacked_symint u0 [-9223372036854775808, 9223372036854775807] at test/inductor/test_torchinductor_dynamic_shapes.py:320 in f (_subclasses/fake_tensor.py:608 in nonzero)
[2024-01-25 08:34:32,994] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] set_replacement u0 = 4
[2024-01-25 08:34:32,995] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] runtime_assert Eq(u0, 4) [guard added] at test/inductor/test_torchinductor_dynamic_shapes.py:321 in f (_decomp/decompositions.py:1229 in split_with_sizes)
[2024-01-25 08:34:32,995] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] set_replacement u0 = 4
[2024-01-25 08:34:33,013] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] create_unbacked_symint u1 [-9223372036854775808, 9223372036854775807] (_subclasses/fake_tensor.py:608 in nonzero)
[2024-01-25 08:34:33,309] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] create_unbacked_symint u2 [-9223372036854775808, 9223372036854775807] (_subclasses/fake_tensor.py:608 in nonzero)
[2024-01-25 08:34:33,331] [0/0] torch.fx.experimental.symbolic_shapes: [INFO] produce_guards
```

The u2 is an unrelated inductor thing. The u1 is from AOTAutograd fake tensor prop; the size of the nonzero tensor never shows up anywhere so we fail to unify it with u0.

Reply:
Ah right. In this case though: we never actually **use** the unbacked symint from the nonzero anywhere. I agree that unifying even in this case would be better than not. I guess I'm curious whether this can actually lead to any problems in practice.

There's now logic in inductor to grab every unbacked symint in the graph (even if they are unused) and do stuff to it like track dependencies, so probably better to be safe than sorry, and replace dup'd symints in all cases.

Reply:
It is true that in this test case, it doesn't actually cause a problem. And I cannot think of an easy way to construct a problematic case. So let's just call this an invariant to be enforced, perhaps for later use.